operators/src/pro/cache/cache_tiles.rs

```diff
@@ -314,23 +309,354 @@ impl<T: Pixel> Stream for CacheTileStream<T> {
                 && tile.time.intersects(&query.time_interval)
             {
                 *idx = i + 1;
-                return std::task::Poll::Ready(Some(Ok(tile.clone())));
+
+                // create a future that decodes the tile
+                let future_data = data.clone();
+                state.set(Some(crate::util::spawn_blocking(move || {
+                    let data = future_data;
+                    let idx = i;
+                    let element = RasterTile2D::<T>::from_stored_element_ref(&data[idx]);
+                    Ok(element)
+                })));
+
+                // maybe the future is already finished, so we can try to return it immediately
+                let tile_future = state
+                    .as_mut()
+                    .as_pin_mut()
+                    .expect("The future must exist since we just set it");
+                // futures ready will return pending if the future is not ready yet
+                let tile_res_res = futures::ready!(tile_future.poll(cx));
+                state.set(None);
+                let Ok(tile_res) = tile_res_res else {
+                return std::task::Poll::Ready(Some(
+                    Err(CacheError::CouldNotDecompressElement),
+                ));
+            };
+                return std::task::Poll::Ready(Some(tile_res));
             }
         }
 
         std::task::Poll::Ready(None)
     }
 }
 
-pub enum TypedCacheTileStream {
-    U8(CacheTileStream<u8>),
-    U16(CacheTileStream<u16>),
-    U32(CacheTileStream<u32>),
-    U64(CacheTileStream<u64>),
-    I8(CacheTileStream<i8>),
-    I16(CacheTileStream<i16>),
-    I32(CacheTileStream<i32>),
-    I64(CacheTileStream<i64>),
-    F32(CacheTileStream<f32>),
-    F64(CacheTileStream<f64>),
+#[derive(Clone, Debug)]
+pub struct CompressedMaskedGrid<D, T> {
+    shape: D,
+    type_marker: PhantomData<T>,
+    data: Vec<u8>,
+    mask: Vec<u8>,
+}
+
+#[derive(Clone, Debug)]
+pub enum CompressedGridOrEmpty<D, T> {
+    Empty(EmptyGrid<D, T>),
+    Compressed(CompressedMaskedGrid<D, T>),
+}
+
+pub type CompressedRasterTile<D, T> = BaseTile<CompressedGridOrEmpty<D, T>>;
+pub type CompressedRasterTile2D<T> = CompressedRasterTile<GridShape2D, T>;
+
+fn unsafe_cast_data_slice_to_u8_slice<T>(data: &[T]) -> &[u8]
+where
+    T: Copy,
+{
+    unsafe { std::slice::from_raw_parts(data.as_ptr().cast::<u8>(), std::mem::size_of_val(data)) }
+}
+
+fn unsafe_cast_u8_slice_to_data_slice<T>(data: &[u8]) -> &[T]
+where
+    T: Copy,
+{
+    unsafe {
+        std::slice::from_raw_parts(
+            data.as_ptr().cast::<T>(),
+            data.len() / std::mem::size_of::<T>(),
+        )
+    }
+}
+
+impl<D, T> CompressedMaskedGrid<D, T> {
+    #[cfg(test)]
+    pub(crate) fn new(shape: D, data: Vec<u8>, mask: Vec<u8>) -> Self {
+        Self {
+            shape,
+            type_marker: PhantomData,
+            data,
+            mask,
+        }
+    }
+
+    pub fn compressed_data_size(&self) -> usize {
```

Comment:
maybe `compressed_data_len` would be less confusing w.r.t to byte size?

Reply:
[4e25ba7](https://github.com/geo-engine/geoengine/pull/848/commits/4e25ba786092b3badcad6b52e91bb75bb020e586)